shell/platform/embedder/embedder_surface_gl_impeller.cc

```diff
@@ -0,0 +1,195 @@
+
+#include "flutter/shell/common/shell_io_manager.h"
+#include "impeller/entity/gles/entity_shaders_gles.h"
+#include "impeller/renderer/backend/gles/context_gles.h"
+#include "impeller/renderer/backend/gles/proc_table_gles.h"
+#include "impeller/scene/shaders/gles/scene_shaders_gles.h"
+
+namespace flutter {
+
+class ReactorWorker final : public impeller::ReactorGLES::Worker {
+ public:
+  ReactorWorker() = default;
+
+  // |ReactorGLES::Worker|
+  bool CanReactorReactOnCurrentThreadNow(
+      const impeller::ReactorGLES& reactor) const override {
+    impeller::ReaderLock lock(mutex_);
+    auto found = reactions_allowed_.find(std::this_thread::get_id());
+    if (found == reactions_allowed_.end()) {
+      return false;
+    }
+    return found->second;
+  }
+
+  void SetReactionsAllowedOnCurrentThread(bool allowed) {
+    impeller::WriterLock lock(mutex_);
+    reactions_allowed_[std::this_thread::get_id()] = allowed;
+  }
+
+ private:
+  mutable impeller::RWMutex mutex_;
+  std::map<std::thread::id, bool> reactions_allowed_ IPLR_GUARDED_BY(mutex_);
+
+  FML_DISALLOW_COPY_AND_ASSIGN(ReactorWorker);
+};
+
+EmbedderSurfaceGLImpeller::EmbedderSurfaceGLImpeller(
+    EmbedderSurfaceGL::GLDispatchTable gl_dispatch_table,
+    bool fbo_reset_after_present,
+    std::shared_ptr<EmbedderExternalViewEmbedder> external_view_embedder)
+    : gl_dispatch_table_(std::move(gl_dispatch_table)),
+      fbo_reset_after_present_(fbo_reset_after_present),
+      external_view_embedder_(std::move(external_view_embedder)),
+      worker_(std::make_shared<ReactorWorker>()) {
+  worker_->SetReactionsAllowedOnCurrentThread(true);
+  // Make sure all required members of the dispatch table are checked.
+  if (!gl_dispatch_table_.gl_make_current_callback ||
+      !gl_dispatch_table_.gl_clear_current_callback ||
+      !gl_dispatch_table_.gl_present_callback ||
+      !gl_dispatch_table_.gl_fbo_callback ||
+      !gl_dispatch_table_.gl_populate_existing_damage ||
+      !gl_dispatch_table_.gl_proc_resolver) {
+    return;
+  }
+  std::vector<std::shared_ptr<fml::Mapping>> shader_mappings = {
+      std::make_shared<fml::NonOwnedMapping>(
+          impeller_entity_shaders_gles_data,
+          impeller_entity_shaders_gles_length),
+      std::make_shared<fml::NonOwnedMapping>(
+          impeller_scene_shaders_gles_data, impeller_scene_shaders_gles_length),
+  };
+  auto gl = std::make_unique<impeller::ProcTableGLES>(
+      gl_dispatch_table_.gl_proc_resolver);
+  if (!gl->IsValid()) {
+    return;
+  }
+
+  impeller_context_ =
+      impeller::ContextGLES::Create(std::move(gl), shader_mappings);
+
+  if (!impeller_context_) {
+    FML_LOG(ERROR) << "Could not create Impeller context.";
+    return;
+  }
+
+  auto worker_id = impeller_context_->AddReactorWorker(worker_);
+  if (!worker_id.has_value()) {
+    FML_LOG(ERROR) << "Could not add reactor worker.";
+    return;
+  }
+
+  FML_LOG(ERROR) << "Using the Impeller rendering backend (OpenGL).";
```

Comment:
Is `ERROR` correct here? I'd expect this to be an `INFO` or `WARNING`

Reply:
ERROR is the only message that shows up across the board and is what we're using for other platforms to disambiguate whether or not impeller is running. We may remove this at some point in the future.

Reply:
Gotcha, marking as resolved